src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java

```diff
@@ -0,0 +1,448 @@
+               continue;
+            }
+         }
+
+         // If a period/frequency limit was set, wait until the next period has started.
+         if (periodLowerLimit > 0.0)
+         {
+            // Uses ThreadTools.parkAtLeast, which will not throw InterruptedException, but
+            // will return early when this thread is interrupted.
+            throttler.waitAndRun(periodLowerLimit);
+         }
+
+         // 1. Clears the interrupt status so we don't run the task with it.
+         // 2. Allows the user to use interrupt to get the throttler to stop waiting.
+         // 3. We won't execute the task if there's nothing scheduled or kill has been called.
+         // 4. Otherwise, we'll go ahead and execute the next task.
+         if (interrupted() && (!running || executionState.getScheduled() == 0L))
+            continue;
+
+         // Run the runTask method, and handle any exception it may throw.
+         executionState.beforeTaskExecution();
+         ExceptionTools.handle(this::runTask, exceptionHandler);
+         executionState.afterTaskExecution();
+      }
+   }
+
+   /** The execution state of the RepeatingTaskThread. */
+   private static class ExecutionState
+   {
+      /**
+       * How many more times to execute the task.
+       *
+       * <ul>
+       *    <li> n > 0 = Task will be executed n more times.
+       *    <li> 0 = Task will not be executed again.
+       *    <li> -1 = Task will be executed repeatedly and indefinitely.
+       */
+      private volatile long scheduledRepetitions = 0L;
+
+      /** Whether the task is currently executing. */
+      private volatile boolean executing = false;
+
+      /** The total number of times the task has completed execution during the lifetime of this thread. */
+      private volatile long completedRepetitions = 0L;
+
+      private synchronized void beforeTaskExecution()
+      {
+         if (scheduledRepetitions > 0)
+            --scheduledRepetitions;
+         executing = true;
+         notifyAll();
+      }
+
+      private synchronized void afterTaskExecution()
+      {
+         executing = false;
+         ++completedRepetitions;
+         notifyAll();
+      }
+
+      private synchronized void setScheduled(long repetitions)
+      {
+         scheduledRepetitions = repetitions;
+         notifyAll();
+      }
+
+      private synchronized void addScheduled(long repetitions)
+      {
+         // If repeating indefinitely, do nothing
+         if (scheduledRepetitions < 0L)
+            return;
+
+         // Add to the scheduled repetition counter, ensuring it doesn't become negative
+         scheduledRepetitions = Math.max(scheduledRepetitions + repetitions, 0L);
+
+         notifyAll();
+      }
+
+      /**
+       * Wait until a change occurs to the thread's execution state.
+       *
+       * @return {@code true} if interrupted.
+       */
+      private synchronized boolean waitForChange()
```

Comment:
Does this need to be synchronized? Seems weird to me

Reply:
Yeah, the call to `wait()` requires it to be synchronized

Reply:
Oh, right. I forgot